File: packages/json/llm-json-stream-ts/packages/llm-json-stream/src/classes/property_delegates/string_property_delegate.ts
/**
 * Delegate for parsing JSON string values.
 * Handles escape sequences and streaming of string content.
 */

import { PropertyDelegate } from "./property_delegate.js";
import { JsonStreamParserController } from "../json_stream_parser.js";

export class StringPropertyDelegate extends PropertyDelegate {
    private buffer = "";
    private isEscaping = false;
    private firstCharacter = true;

    constructor(
        propertyPath: string,
        parserController: JsonStreamParserController,
        onComplete?: () => void,
    ) {
        super(propertyPath, parserController, onComplete);
    }

    addCharacter(character: string): void {
        // Handle opening quote
        if (this.firstCharacter && character === '"') {
            this.firstCharacter = false;
            return;
        }
        if (this.firstCharacter) {
            throw new Error(
                `StringPropertyDelegate expected starting quote but got: ${character}`,
            );
        }

        // Handle escape sequences
        if (this.isEscaping) {
            switch (character) {
                case '"':
                    this.buffer += '"';
                    break;
                case "\\":
                    this.buffer += "\\";
                    break;
                case "/":
                    this.buffer += "/";
                    break;
                case "b":
                    this.buffer += "\b";
                    break;
                case "f":
                    this.buffer += "\f";
                    break;
                case "n":
                    this.buffer += "\n";
                    break;
                case "r":
                    this.buffer += "\r";
                    break;
                case "t":
                    this.buffer += "\t";
                    break;
                default:
                    // For unknown escape sequences, include both backslash and character
                    this.buffer += "\\";
                    this.buffer += character;
                    break;
            }
            this.isEscaping = false;
            return;
        }

        // Start escape sequence
        if (character === "\\") {
            this.isEscaping = true;
            return;
        }

        // End of string - closing quote
        if (character === '"') {
            this.isDone = true;
            // Emit final chunk if there's any remaining buffer
            if (this.buffer.length > 0) {
                this.addPropertyChunk(this.buffer);
                this.buffer = "";
            }
            // Complete the string controller
            this.parserController.completeProperty(this.propertyPath, "");
            this.onComplete?.();
            return;
        }

        // Regular character - add to buffer
        this.buffer += character;
    }

    override onChunkEnd(): void {
        if (this.buffer.length === 0 || this.isDone) return;

        // Emit accumulated buffer content
        this.addPropertyChunk(this.buffer);
        this.buffer = "";
    }
}

File: packages/json/llm-json-stream-python/src/llm_json_stream/property_delegates/string_property_delegate.py
"""Delegate for parsing JSON string values."""

from __future__ import annotations

from ..property_delegate import PropertyDelegate


class StringPropertyDelegate(PropertyDelegate):
    def __init__(self, property_path: str, parser_controller, on_complete=None) -> None:
        super().__init__(property_path, parser_controller, on_complete)
        self._buffer = ""
        self._is_escaping = False
        self._first_character = True

    def add_character(self, character: str) -> None:
        if self._first_character and character == '"':
            self._first_character = False
            return
        if self._first_character:
            raise ValueError(
                f"StringPropertyDelegate expected opening quote, got: {character}"
            )

        if self._is_escaping:
            escape_map = {
                '"': '"',
                "\\": "\\",
                "/": "/",
                "b": "\b",
                "f": "\f",
                "n": "\n",
                "r": "\r",
                "t": "\t",
            }
            self._buffer += escape_map.get(character, f"\\{character}")
            self._is_escaping = False
            return

        if character == "\\":
            self._is_escaping = True
            return

        if character == '"':
            self._is_done = True
            if self._buffer:
                self.add_property_chunk(self._buffer)
                self._buffer = ""
            self._parser_controller.complete_property(self._property_path, "")
            if self._on_complete:
                self._on_complete()
            return

        self._buffer += character

    def on_chunk_end(self) -> None:
        if not self._buffer or self._is_done:
            return
        self.add_property_chunk(self._buffer)
        self._buffer = ""
